ml.q

```diff
@@ -258,12 +258,18 @@ cgroup:{[df;X;C]value group imin f2nd[df X] C}
 / matri(X) to the nearest (C)entroid and then uses the (c)entroid (f)unction
 / to update the centroid location.
 lloyd:{[df;cf;X;C]cf X@\: cgroup[df;X;C]}
+/ uses (r)esponsibility (f)unction David Mackay's Information Theory... pg 289
+lloyds:{[df;cf;rf;X;C] cf[rf .ml.f2nd[df X] C;X]} /soft assignment 
 
 kmeans:lloyd[edist2;avg'']      / k-means
 kmedians:lloyd[mdist;med'']     / k-medians
 khmeans:lloyd[edist2;hmean'']   / k harmonic means
 skmeans:lloyd[cosdist;normalize (avg'')::] / spherical k-means
 
+kmeanss:lloyds[edist2;wavg\:/:;{x=\:min x}] /k-means using loyd with rf
+/ kmeansoft1 David Mackay (b)eta stiffness param
```

Comment:
i'd prefer a more descriptive function qualifier than a trailing 1.  what does the 1 represent?

Reply:
agree, but mackay develops two more versions of soft kmeans in chapter 22, which I haven't had time to implement yet.